### PR TITLE
Fix for disappearing fields in tables that are ajax loaded.

### PR DESCRIPTION
**Pull request by spficklin**
# Bug Fix

Issue #805 

## Description

When hiding of empty fields for a content type is turned on, then fields that are marked to not auto attach (meaning they should be loaded via AJAX) are automatically removed from the Tripal Pane table. This is because they are technically "empty" as data has not yet been added to them (it gets added via AJAX). This is incorrect behavior. They should not be removed from the table until after their data is loaded. The tripal_field_group_table_rows_alter function is the culprit.

## Testing?

To test that there is a problem
1.  Do not pull this branch just yet
2.  Make sure you have checked from the mRNA content type the hide empty fields and the load via ajax settings.
3. Go to any mRNA page that has a sequence.  I use the Citrus sinensis data from the Tripal tutorial.  
4. Click the 'Reload' tab and the page should reload.  The mRNA sequence is missing!
5. Turn off the 'hide empty fields' setting for the mRNA content type, return to the mRNA page and then click the 'Reload' tab. You should now see the sequence.

To test the fix:
1.  Pull the branch
2.  Perform steps 1-4 above.  This time however, the mRNA sequence is present.

### CODE REVIEW

File: tripal/includes/TripalEntityController.inc [hunk 1]
```diff
@@ -688,14 +688,19 @@ class TripalEntityController extends EntityAPIController {
           // attach then we will ignore it. It can only be set by providing
           // the id in the $field_id array handled previously.
           else {
-            
-            // Do not load fields that are not auto attached.  Instead set
-            // their value to an empty string and set the #processed key to
-            // FALSE.
+
+            // Does the field instance have an 'auto_attach' setting?
+            $auto_attach = FALSE;
```
Comment (spficklin) [inline tripal/includes/TripalEntityController.inc:693]: Just a note... I changed these lines of code to hopefully prevent problems in the future if a developer forgets to set the 'auto_attach'  we need it to have a default.

File: tripal/includes/tripal.user.inc [hunk 2]
```diff
@@ -3,47 +3,47 @@
...
+
   // Get all of the files that have been uploaded by the user.
   $sql = "
      SELECT FM.fid, FM.filename, FM.uri, FM.uid, FM.filesize, TGEF.expiration_date
      FROM {file_managed} FM
        INNER JOIN {file_usage} FU on FM.fid = FU.fid and FM.uid = :user_id
        LEFT JOIN {tripal_expiration_files} TGEF on TGEF.fid = FU.fid
      WHERE FU.module = 'tripal'
-     ORDER BY FM.filename
+     ORDER BY FM.uri
```
Comment (spficklin) [inline tripal/includes/tripal.user.inc:14]: Whoops. I didn't mean for this fix to make it in to this PR.  But it's so simple I'll leave it.  It just fixes a sorting issue with the user's uploaded file list.

File: tripal/tripal.module [hunk 3]
```diff
@@ -1565,18 +1551,18 @@ function tripal_field_group_table_rows_alter(&$element, &$children) {
       $bundle = $element[$child]['#bundle'];
       $bundle_info = tripal_load_bundle_entity(array('name' => $bundle));
 
+      $processed = FALSE;
+      if (array_key_exists('#processed', $element[$child]['#object']->$child)) {
+        $processed = $element[$child]['#object']->$child['#processed'];
+      }
+
```
Comment (spficklin) [inline tripal/tripal.module:1558]: This set of code is the beginning of the true fix for the problem.

**Comment by spficklin** [comment]
Sorry for all of the formatting changes in this PR.  I turned on automatic removal of white space in my IDE.  Also, I fixed a few other field formatters (e.g. publication) that weren't properly dealing with empty values.  They don't affect this bug, but it took me a while to track down the problem and I wound up fixing these things as well.

**Comment by laceysanderson** [comment]
I'm having difficulties reproducing this problem...

## On 7.x-3.x (f543712afd3b21860d029fbf3d9dc69c112ede76)
 - [x] Make sure you have checked from the mRNA content type the hide empty fields and the load via ajax settings.
 - [x] Go to any mRNA page that has a sequence. **I created a new mRNA through the UI with dummy nucleotide sequence**
- [ ] Click the 'Reload' tab and the page should reload. **I see the sequence, even after reloading and clearing the full cache multiple times**
- [x] Turn off the 'hide empty fields' setting for the mRNA content type, return to the mRNA page and then click the 'Reload' tab. You should now see the sequence.
screen shot 2018-12-31 at 10 37 02 am

**Comment by spficklin** [comment]
@laceysanderson it may be a timing issue and the order that the `tripal_field_group_table_rows_alter` function runs and when the load function gets called for the field.    If so you might not be able to reproduce the problem.

**Comment by bradfordcondon** [comment]
I also cant reproduce the problem.

**Comment by bradfordcondon** [comment]
on master with hide empty fields

screen shot 2019-01-03 at 8 46 38 am

on your branch with hide empty fields:

screen shot 2019-01-03 at 8 47 42 am


So im able to reproduce the problem and confirm the fix resolves it.  approved.

**Review by bradfordcondon — APPROVED** [review]

**Review by laceysanderson — APPROVED** [review]
Did a code review and it all looks reasonable and makes sense 👍

**Comment by laceysanderson** [comment]
Since we have two reviews including one functional review, I'm merging this! Thanks for the fix @spficklin

**Comment by spficklin** [comment]
Thanks for the reviews!

**Pull request merged**